### PR TITLE
refactor: Operations

### PR DESCRIPTION
**Pull request by shubham-cmyk**
make code much more clear and readable format.

### CODE REVIEW

File: pkg/runner/operations/delete/delete.go [hunk 1]
```diff
@@ -29,50 +29,64 @@ func New(client client.Client, obj ctrlclient.Object, namespacer namespacer.Name
...
-			}
-			return false, nil
-		}); err != nil {
+		if err := o.waitForDeletion(ctx, &candidate); err != nil {
+			logger.Log(logging.Delete, logging.ErrorStatus, color.BoldRed, logging.ErrSection(err))
 			return err
 		}
 	}
+
+	logger.Log(logging.Delete, logging.DoneStatus, color.BoldGreen)
```
Comment (eddycharly) [inline pkg/runner/operations/delete/delete.go:69]: things like this are not correct, it should be in a defer to guarantee the log is printed

**Comment by eddycharly** [comment]
I'll review this in depth during the weekend, this is easy to make a mistake.

**Comment by shubham-cmyk** [comment]
Closed by mistake

File: pkg/runner/operations/apply/apply.go [hunk 2]
```diff
@@ -37,74 +37,89 @@ func New(client client.Client, obj ctrlclient.Object, namespacer namespacer.Name
...
-		if _err == nil {
-			logger.Log(logging.Apply, logging.DoneStatus, color.BoldGreen)
+		if err != nil {
+			logger.Log(logging.Apply, logging.ErrorStatus, color.BoldRed, logging.ErrSection(err))
 		} else {
-			logger.Log(logging.Apply, logging.ErrorStatus, color.BoldRed, logging.ErrSection(_err))
+			logger.Log(logging.Apply, logging.DoneStatus, color.BoldGreen)
 		}
 	}()
+
```
Comment (eddycharly) [inline pkg/runner/operations/apply/apply.go:49]: i'm not a big fan of blank lines 🤷

File: pkg/runner/operations/apply/apply.go [hunk 3]
```diff
@@ -37,74 +37,89 @@ func New(client client.Client, obj ctrlclient.Object, namespacer namespacer.Name
...
+func (o *operation) applyResource(ctx context.Context, logger logging.Logger) error {
+	err := wait.PollUntilContextCancel(ctx, internal.PollInterval, false, func(ctx context.Context) (bool, error) {
+		err := o.tryApplyResource(ctx)
+		return err == nil, err
 	})
+	if err != nil {
+		return err
+	}
+
+	return nil
```
Comment (eddycharly) [inline pkg/runner/operations/apply/apply.go:69]:
```suggestion
	return wait.PollUntilContextCancel(ctx, internal.PollInterval, false, func(ctx context.Context) (bool, error) {
		err := o.tryApplyResource(ctx)
		return err == nil, err
	})
```

File: pkg/runner/operations/apply/apply.go [hunk 4]
```diff
@@ -37,74 +37,89 @@ func New(client client.Client, obj ctrlclient.Object, namespacer namespacer.Name
...
+}
+
+func (o *operation) tryApplyResource(ctx context.Context) error {
+	var actual unstructured.Unstructured
+	actual.SetGroupVersionKind(o.obj.GetObjectKind().GroupVersionKind())
+	err := o.client.Get(ctx, client.ObjectKey(o.obj), &actual)
+	if err == nil {
+		return o.updateResource(ctx, &actual)
+	} else if kerrors.IsNotFound(err) {
```
Comment (eddycharly) [inline pkg/runner/operations/apply/apply.go:78]: no need for the else here

File: pkg/runner/operations/apply/apply.go [hunk 5]
```diff
@@ -37,74 +37,84 @@ func New(client client.Client, obj ctrlclient.Object, namespacer namespacer.Name
...
+		if err != nil {
+			logger.Log(logging.Apply, logging.ErrorStatus, color.BoldRed, logging.ErrSection(err))
 		} else {
-			logger.Log(logging.Apply, logging.ErrorStatus, color.BoldRed, logging.ErrSection(_err))
+			logger.Log(logging.Apply, logging.DoneStatus, color.BoldGreen)
 		}
 	}()
 	if o.namespacer != nil {
-		if err := o.namespacer.Apply(o.obj); err != nil {
+		if err = o.namespacer.Apply(o.obj); err != nil {
```
Comment (eddycharly) [inline pkg/runner/operations/apply/apply.go:50]:
```suggestion
		if err := o.namespacer.Apply(o.obj); err != nil {
```

File: pkg/runner/operations/create/create.go [hunk 6]
```diff
@@ -36,50 +36,74 @@ func New(client client.Client, obj ctrlclient.Object, namespacer namespacer.Name
...
-		if _err == nil {
-			logger.Log(logging.Create, logging.DoneStatus, color.BoldGreen)
+		if err != nil {
+			logger.Log(logging.Create, logging.ErrorStatus, color.BoldRed, logging.ErrSection(err))
 		} else {
-			logger.Log(logging.Create, logging.ErrorStatus, color.BoldRed, logging.ErrSection(_err))
+			logger.Log(logging.Create, logging.DoneStatus, color.BoldGreen)
 		}
 	}()
+
```
Comment (eddycharly) [inline pkg/runner/operations/create/create.go:48]: blank lines

File: pkg/runner/operations/delete/delete.go [hunk 7]
```diff
@@ -29,50 +29,65 @@ func New(client client.Client, obj ctrlclient.Object, namespacer namespacer.Name
...
+	candidates, err := internal.Read(ctx, o.obj, o.client)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
 			return nil
 		}
-		return _err
+		return err
 	}
+
 	for i := range candidates {
```
Comment (eddycharly) [inline pkg/runner/operations/delete/delete.go:61]: it would be better to call all delete then wait for all deletions.

Comment (shubham-cmyk) [reply pkg/runner/operations/delete/delete.go:61]: right it would reduce the wait time for the deletion

**Review by eddycharly — APPROVED** [review]

**Review by eddycharly — APPROVED** [review]

**Pull request merged**